### PR TITLE
Draft to implement upgrade state/disable services

### PR DESCRIPTION
**Pull request by nkrinner**
Add a role to the updater barclamp that allows to disable
all openstack services and crowbar_join and disables and
stops chef-client

### CODE REVIEW

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+if node[:platform] == "suse"
+  service "chef-client" do
+    action [:disable, :stop]
+  end
+end
+
+### Transition node to 'upgrade' status
+node[:state] = "upgrade"
+node.save
+
```
Comment (nkrinner) [inline chef/cookbooks/updater/recipes/upgrade.rb]: wip

File: chef/cookbooks/updater/recipes/default.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #
-# Cookbook Name:: updater
+ Cookbook Name:: updater
```
Comment (jsuchome) [inline chef/cookbooks/updater/recipes/default.rb]: Delete key pressed by chance I guess...

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+if !node[:updater].has_key?(:upgrade_one_shot_run) || !node[:updater][:upgrade_one_shot_run]
+
+  node[:updater][:upgrade_one_shot_run] = true
+  node.save
+
+
+  if node[:platform] == "suse"
+    bash "disable_openstack_services" do
+      code <<-EOF
+        for i in /etc/init.d/openstack-* /etc/init.d/openvswitch-switch /etc/init.d/ovs-usurp-config-* /etc/init.d/drbd /etc/init.d/openais;
```
Comment (rhafer) [inline chef/cookbooks/updater/recipes/upgrade.rb]: You might want to have a look at: https://etherpad.nue.suse.com/p/cloud_5_to_6_controller_upgrade for a more comprehensive (but maybe still incomplete) list of services to disable.

BTW, I am not sure if we really can have openvswitch-switch disabled in all possible scenarios, but for now that should be fine.

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          fi
+        done
+      EOF
+    end
+  end
+
+  # Disable crowbar-join
+  if node[:platform] == "suse"
+    service "crowbar_join" do
+      action :disable
```
Comment (tboerger) [inline chef/cookbooks/updater/recipes/upgrade.rb]: `only_if` block

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    end
+  end
+
+  # Disable chef-client
+  if node[:platform] == "suse"
+    service "chef-client" do
+      action [:disable, :stop]
```
Comment (tboerger) [inline chef/cookbooks/updater/recipes/upgrade.rb]: `only_if` block

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+# is rather complicated. So instead we fall back to a simple bash hack
+
+if !node[:updater].has_key?(:upgrade_one_shot_run) || !node[:updater][:upgrade_one_shot_run]
+
+  node[:updater][:upgrade_one_shot_run] = true
+  node.save
+
+
+  if node[:platform] == "suse"
+    bash "disable_openstack_services" do
```
Comment (tboerger) [inline chef/cookbooks/updater/recipes/upgrade.rb]: `only_if` block

File: crowbar_framework/app/models/updater_service.rb [hunk 7]
```diff
@@ -64,6 +72,17 @@ def apply_role_pre_chef_call(old_role, role, all_nodes)
         node.save
       end
     end
+
+    # Remove [:updater][:upgrade_one_shot_run] flag from node
+    nodes = NodeObject.find("roles:updater-upgrade") || []
```
Comment (vuntz) [inline crowbar_framework/app/models/updater_service.rb]: Can't you get that info by looking at the role?

File: crowbar_framework/app/models/updater_service.rb [hunk 8]
```diff
@@ -73,6 +92,8 @@ def apply_role_pre_chef_call(old_role, role, all_nodes)
    #  end
    #end
     @logger.debug("Updater apply_role_post_chef_call: exiting")
+
+
```
Comment (vuntz) [inline crowbar_framework/app/models/updater_service.rb]: Drop empty lines?

File: crowbar.yml [hunk 9]
```diff
@@ -26,6 +26,6 @@ barclamp:
 crowbar:
   layout: 1
   order: 99
-  run_order: 99
-  chef_order: 99
+  run_order: 2000
+  chef_order: 2000
```
Comment (vuntz) [inline crowbar.yml:30]: I guess you're doing that to disable the services and not have them restarted by later recipes. That won't work for anything that happens in notifications, though.

The right "fix" here is that all the other (at least, the openstack ones) cookbooks should stop being applied. Not sure how to best do this.

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  # Disable chef-client
+  if node[:platform] == "suse"
+    service "chef-client" do
+      action [:disable, :stop]
+    end
+  end
+
+  ### Transition node to 'upgrade' state
+  node[:state] = "upgrade"
+  node.save
```
Comment (vuntz) [inline chef/cookbooks/updater/recipes/upgrade.rb]: This should definitely not be done in the cookbook, but it should be done in the model in the rails app. You can use the apply_role_post_chef_call method for this.

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+# Disable openstack services
```
Comment (jsuchome) [inline chef/cookbooks/updater/recipes/upgrade.rb:21]: Maybe move this comment to actual block that does the disabling?

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+# We don't know which openstack services are enabled on the node and
+# collecting that information via the attributes provided by chef is
+# is rather complicated. So instead we fall back to a simple bash hack
+
+if !node[:updater].has_key?(:upgrade_one_shot_run) || !node[:updater][:upgrade_one_shot_run]
+
+  node[:updater][:upgrade_one_shot_run] = true
```
Comment (jsuchome) [inline chef/cookbooks/updater/recipes/upgrade.rb:28]: Is this variable really needed? Can't you check for the node state, if it is already "upgrade"?
(Maybe I'm missing something)

Comment (nkrinner) [inline chef/cookbooks/updater/recipes/upgrade.rb:28]: Good point, I guess we can drop that once the state manipulation works fine.

**Comment by rhafer** [comment]
Hm, I might not yet understand the whole upgrade procedure for Tex->Master completely. But to me it sounds really hard to implement a catch all recipe that disables everything correctly for every supported configuration out there. 

Wouldn't it be a lot cleaner to implement this similar to how the remove roles are supposed to work? I.e. have an specific disable role for all "normal" roles (e.g. disable-neutron-server, disable-keystone-server, ...) where the runlist of those roles "just" contains a single recipe that just disables the service for a specific barclamp.
(And then adjust the "elements_states" in the deployment section for all barclamp to add those disable roles to the runlist only for the new maintainance/upgrade state).
This is just my naive opinion from a very far distance and I might be completely wrong on this (because I don't know all the details). I just have the fear that by trying to implement everything in a single recipe in the update barclamp we're going to endup with a gigantic hack.

**Comment by nkrinner** [comment]
@rhafer I also looked at the work that has been done before to disable roles/services and tried to find a solution based on that. But in my approach it turned out to be even more hacky and required more hardcoding than the present solution does. 

As I understand it, with every solution that comes with roles, the ordering in the run_list will remain a problem. But I am also currently looking for a better solution right now and I am checking the crowbar-core code to see where best to manipulate the state of services on the nodes.

**Comment by nkrinner** [comment]
After looking at the code a bit longer and thinking about it, a valid approach could be to drop nodes active role and replace it with a role that only hast the updater-upgrade role in its run list. This would happen within apply_chef_role_before_chef. That would allow us to preserve the old role and avoid the interference of unwanted roles on the run list.

**Comment by nkrinner** [comment]
Added a reworked draft. Untested and most likely broken at the moment, but the concept should be visible now and seems to me like the cleaner solution - if I manage to get this working.

File: crowbar_framework/app/models/updater_service.rb [hunk 13]
```diff
@@ -55,15 +63,37 @@ def create_proposal
...
+      node = NodeObject.find_node_by_name n
+      if node.role? "updater-upgrade"
+        storage_role = role
+        # TODO: Make sure that storage_role.name does not already exist
+        storage_role.name= role.name+"-upgrade_storage"
+        storage_role.save
+        node.delete_from_run_list(role.name)
+        # TODO: Find out if there can be more than one role on the run_list -
+        # If so, handle that in a proper way
+        node.add_to_run_list("updater-upgrade", 2000)
```
Comment (jsuchome) [inline crowbar_framework/app/models/updater_service.rb:94]: We don't really need a high priority here, do we? There should be just one role in this run list I assume

File: crowbar_framework/app/models/updater_service.rb [hunk 14]
```diff
@@ -75,6 +105,30 @@ def apply_role_pre_chef_call(old_role, role, all_nodes)
...
+      if node.role? "updater-upgrade"
+        # Put together storage role name
+        role_name = NodeObject.make_role_name(node.name)
+        role  = RoleObject.find_role_by_name(role_name+"-upgrade_storage")
+        role.name = role_name
+        node.delete_from_run_list("updater-upgrade")
+#        priority = runlist_priority_map[item] || local_chef_order
+        priority = 1
+        # FIXME: Not sure if this works, and if it does, find solution for priority
+        node.add_to_run_list(role.name, priority)
```
Comment (jsuchome) [inline crowbar_framework/app/models/updater_service.rb:122]: Shouldn't this be enough?

```
node.run_list.run_list_items.clear
node.run_list.run_list_items << make_role_name(node.name)
```

File: chef/cookbooks/updater/recipes/upgrade.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+      do
+        if test -e $i
+        then
+          initscript=`basename $i`
+          insserv -r $initscript
+        fi
+      done
+    EOF
+    only_if { node[:platform] == "suse" }
+  end
```
Comment (jsuchome) [inline chef/cookbooks/updater/recipes/upgrade.rb:44]: HA parts are needed as well: 
- disable corosync

(code for this should be in the 4-5 upgrade script)

**Comment by nkrinner** [comment]
Depends on https://github.com/crowbar/barclamp-crowbar/pull/1357

File: crowbar_framework/app/models/updater_service.rb [hunk 16]
```diff
@@ -30,12 +30,20 @@ def role_constraints
...
+          "exclude_platform" => {
+            "windows" => "/.*/"
+          }
         }
       }
     end
   end
 
-  def create_proposal
+  def create_proposal(role_name=nil)
```
Comment (jsuchome) [inline crowbar_framework/app/models/updater_service.rb]: Why not `updater` ?

File: crowbar_framework/app/controllers/updater_controller.rb [hunk 17]
```diff
@@ -15,6 +15,17 @@
...
+  before_filter :initialize_service
+  before_filter :controller_to_barclamp
+
+  def controller_to_barclamp
+    @bc_name = "updater"
+    @service_object.bc_name = @bc_name
+  end
+
+  def prepare_upgrade
+    UpdaterService.prepare_upgrade
```
Comment (tboerger) [inline crowbar_framework/app/controllers/updater_controller.rb]: Is this an action? Than provide proper `respond_to` blocks please.

File: crowbar_framework/app/controllers/updater_controller.rb [hunk 18]
```diff
@@ -15,6 +15,17 @@
 #
 
 class UpdaterController < BarclampController
+  before_filter :initialize_service
+  before_filter :controller_to_barclamp
+
+  def controller_to_barclamp
```
Comment (tboerger) [inline crowbar_framework/app/controllers/updater_controller.rb]: You don't need that. You are in the `UpdaterController` and it initializes the `UpdaterService` within `initialize_service`

File: crowbar_framework/app/controllers/updater_controller.rb (hunk 17 above)
Comment (tboerger) [inline crowbar_framework/app/controllers/updater_controller.rb]: And please use the `@service_object` instance :)

File: crowbar_framework/app/models/updater_service.rb [hunk 19]
```diff
@@ -44,26 +52,67 @@ def create_proposal
...
+    NodeObject.all.each do |n|
+      unless n.nil? || n.admin?
+        base["deployment"]["updater"]["elements"] = {
+        "updater-upgrade" => nodes.select { |x| not ["discovering", "discovered"].include?(x.status) }.map { |x| x.name }
+        }       
+      end  
+    end
+
+    #Apply the barclamp
+    ServiceObject.proposal_commit params
```
Comment (tboerger) [inline crowbar_framework/app/models/updater_service.rb]: `self.class.proposal_commit params`

File: crowbar_framework/config/locales/updater/en.yml [hunk 20]
```diff
@@ -15,6 +15,9 @@
 #
 
 en:
+  nav:
+    utils:
+      upgrade: 'Prepare nodes for upgrade'
```
Comment (tboerger) [inline crowbar_framework/config/locales/updater/en.yml]: This title should be shorter.

File: crowbar.yml [hunk 21]
```diff
@@ -26,6 +26,13 @@ barclamp:
...
   proposal_schema_version: 3
+
+nav:
+  utils:
+    upgrade:
+      order: 50
+      route: 'upgrade_path'
```
Comment (tboerger) [inline crowbar.yml]: As this action updates the proposals this should be a post action with a confirmation.

File: crowbar_framework/app/models/updater_service.rb [hunk 22]
```diff
@@ -75,6 +124,20 @@ def apply_role_pre_chef_call(old_role, role, all_nodes)
...
   end
 
+  def apply_role_post_chef_call(old_role, role, all_nodes)
+    all_nodes.each do |n|
+      node = NodeObject.find_node_by_name n
+      if node.role? "updater-upgrade"
+        role_name = NodeObject.make_role_name(node.name)
+        node.run_list.run_list_items.clear
+        node.run_list.run_list_items << "role[#{role_name}]"
+        node.crowbar["state"] = "upgrade"
```
Comment (vuntz) [inline crowbar_framework/app/models/updater_service.rb:123]: Thinking out loud: I'm thinking we want to split that upgrade state in two: crowbar_upgrade and os_upgrade. What we're doing in this PR would be crowbar_upgrade, and when crowbar is then upgraded and the user says "okay, I want to now upgrade all nodes too", we move to "os_upgrade" which triggers autoyast profile generation, reboot of nodes, etc.

**Comment by nkrinner** [comment]
Closing, as I won't continue working on this pr and someone else will take over.

**Pull request closed without merging**